File: js/potoo/src/mqtt.ts
export type Topic = string

export interface Client {
    connect: (config: ConnectConfig) => Promise<void>,
    publish: (message: Message)      => void,
    subscribe: (filter: Topic)       => Promise<void>,
    unsubscribe: (filter: Topic)     => Promise<void>,
}

export interface ConnectConfig {
    on_disconnect:   () => void,
    on_message:      (message: Message) => void,
    will_message?:    Message,
}

export interface Message {
    topic: Topic,
    payload: string,
    retain: boolean,
}

export function join_topic_list(topics: Array<Topic>): Topic {
    return topics.reduce(join_topics, '')
}

export function join_topics(a: Topic, b: Topic): Topic {
    a = trim_topic(a)
    b = trim_topic(b)
    if (a == '') return b;
    if (b == '') return a;
    return a + '/' + b
}

export function trim_topic(t: Topic): Topic {
    return t.replace(/^\/+|\/+$/g, '')
}

export function strip_topic(prefix: Topic, t: Topic): Topic | null {
    prefix = trim_topic(prefix)
    t = trim_topic(t)
    if (prefix == t) {
        return ''
    }
    prefix += '/'
    if (t.startsWith(prefix)) {
        return t.substring(prefix.length)
    }
    return null
}

File: go/potoo/mqtt/mqtt.go
package mqtt

type Topic []byte

type ConnectConfig struct {
	OnDisconnect chan<- error
	OnMessage    chan<- Message
	WillMessage  Message
}

type Message struct {
	Topic   Topic
	Payload []byte
	Retain  bool
}

type Client interface {
	// TODO: decide on syncness/asyncness of each of these.
	// or maybe the implementation should decide?
	Connect(config *ConnectConfig) error
	Publish(message Message)
	Subscribe(filter Topic) // must be synchronous
	Unsubscribe(filter Topic)
	DisconnectWithWill()
	Disconnect()
}

func (c *ConnectConfig) Copy() *ConnectConfig {
	msg := c.WillMessage.Copy()
	return &ConnectConfig{
		OnDisconnect: c.OnDisconnect,
		OnMessage:    c.OnMessage,
		WillMessage:  *msg,
	}
}

func (m *Message) Copy() *Message {
	return &Message{
		Topic:   append(Topic{}, m.Topic...),
		Payload: append([]byte{}, m.Payload...),
		Retain:  m.Retain,
	}
}
